audonnx/core/model.py

```diff
@@ -347,6 +365,64 @@ def _device_to_providers(
     return providers
 
 
+def _concat(
+        y: typing.Dict[str, np.ndarray],
+        shapes: typing.Sequence[typing.List[int]],
+):
+    r"""Flatten dictionary by concatenating values."""
+
+    y = list(y.values())
+
+    # special case if all shapes are [-1]
+    if all(map([-1].__eq__, shapes)):
+        return np.stack(y)
+
+    axis = _concat_axis(shapes)
+    if axis == -1:
+        raise RuntimeError(
+            f'To concatenate outputs '
+            f'number of dimensions, '
+            f'position of dynamic axis, '
+            f'and all dimensions except the last non-dynamic axis '
+            f'must match. '
+            f'This does not apply to: '
+            f'{shapes}'
+        )
+
+    return np.concatenate(y, axis=axis)
+
+
+def _concat_axis(shapes: typing.Sequence[int]) -> int:
+    r"""Return concat dimension or -1 if not possible."""
+
+    # number of dimensions do not match
+    if not len(set(map(len, shapes))) == 1:
+        return -1
+
+    # dynamic axis in different positions
+    if not len(set(map(_dynamic_axis, shapes))) == 1:
+        return -1
+
+    axis = len(shapes[0]) - (2 if shapes[0][-1] == -1 else 1)
+
+    # remove concat axis
+    shapes_wo_axis = [shape[:axis] + shape[axis + 1:] for shape in shapes]
+
+    # non-concat dimensions do not match
+    if not all(map(shapes_wo_axis[0].__eq__, shapes_wo_axis)):
+        return -1
+
+    return axis
+
+
+def _dynamic_axis(shape: typing.Sequence[int]) -> int:
+    r"""Return dimension of dynamic axis or -1 if none."""
+    for idx, dim in enumerate(shape):
+        if dim == -1:
+            return idx
+    return -1
```

Comment:
It makes the code slightly harder to understand that we encode the dynamic axis with `-1`, but also if a search does not succeed. Maybe we could use `None` in the later case?

Reply:
Applied your suggestion.